Linter: check for string changes without new IDs

bcolsson opened the pull request:
Revised PR based on feedback from #615
- Now uses compare-locales to parse
- Bash script no longer changes filenames, instead moves files to old/ and new/ folders
- Integrated into reference locale linter action

flodolo commented on .github/scripts/detect_unchanged_ids.py:
```diff
@@ -0,0 +1,107 @@
+
+def main():
+    arg_parser = argparse.ArgumentParser()
+    arg_parser.add_argument("dir_old")
```
Please make these named arguments and add some description, at least it's impossible to swap the files by mistake, e.g.

```
    parser.add_argument(
        "--old",
        required=True,
        help="Folder with the previous version of the file",
    )
```

bcolsson replied:
Fixed

flodolo commented on .github/scripts/detect_unchanged_ids.py:
```diff
@@ -0,0 +1,107 @@
+
+
+    for file_old, file_new in zip(files_old, files_new):
+        filename = file_old.replace(f"{path_old}{os.sep}", "")
```
Unless I'm misunderstanding the intent here:

```suggestion
        filename = os.path.relpath(file_old, path_old)
```

bcolsson replied:
I was looking for something in the path library that would do it, but didn't quite figure out that relpath would do it. Thanks!

flodolo commented on .github/scripts/detect_unchanged_ids.py:
```diff
@@ -0,0 +1,107 @@
+            for id in ids:
+                output.append(
+                    f"\nID: {id}" 
+                    f"\nBefore: {errors[file][id]['before']}" 
```
Maybe Previous/New instead of Before/After?

flodolo commented on .github/scripts/detect_unchanged_ids.py:
```diff
@@ -0,0 +1,107 @@
+        output.append(f"\nTotal errors: {total_errors}")
+
+        print("\n".join(output))
+        sys.exit(1)
```
This should work without a separate print.

```suggestion
        sys.exit("\n".join(output))
```

bcolsson replied:
Would we not need an exit code to indicate the failure?

flodolo replied:
Anything but zero would do, so existing with the string has the same effect.

flodolo reviewed (changes requested):
This looks good, although I haven't tested it yet.

While this is triggered only for changes to `locale/en`, it will still analyze (and fail) for other locales. We need to limit checks to the reference locale by adding an explicit argument.

flodolo commented:
Forgot one note: don't forget to run `black` on the script.

flodolo commented:
Couple of things I forgot:
* Can you rebase against the current default branch? There were a few changes in workflows
* Run black on the script (I could spot a double empty line)

bcolsson commented:
I messed up the rebase pretty badly, so diff isn't saved, but made the following changes:
- Renamed folders/ updated python script call in workflow
- Implemented all your feedback

As I write this I remember I forgot to fix the issue where we need to restrict to only reference files, so that change is incoming.

flodolo commented on .github/scripts/detect_unchanged_ids.py:
```diff
@@ -0,0 +1,116 @@
+    sys.exit(1)
+
+
+def get_file_list(path, reference_locale="locale/en"):
```
The goal would be to be able to reuse this in other places, so it needs to be a parameter that you pass to the script, and also an actual locale code (`en` in this case)

flodolo replied:
On second thoughts, I see how using a locale code might make things harder, so we can pass the "path to reference locale" as argument.

flodolo commented on .github/workflows/reference_linter.yaml:
```diff
@@ -30,3 +32,13 @@ jobs:
+        run: |
+            diff=$( git diff --name-only ${{ github.event.pull_request.base.sha }} $GITHUB_SHA )
+            for file in $diff; do
+              mkdir -p "old/${file%/*}/" && mkdir -p "new/${file%/*}/"
```
You missed updating this line, so the workflow is currently broken.

flodolo reviewed (changes requested):
You lost the change to requirements for compare-locales, so the workflow is currently not running. 

I suggest to set it up on master in your fork for testing, then you can create pull requests against your own fork to make sure it's working.

Heads-up: always double check that you're making PRs against your fork, the default will be to use this repo, and I've made that mistake before (as recently as last week…)

flodolo reviewed (approved):
Looks good now 👍🏼
